### PR TITLE
WIP: Initial CLE integration to a CLEImage class

### PR DESCRIPTION
**Pull request by aaron-bray**

### CODE REVIEW

File: include/itkCLEDataManager.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+/** \class CLEDataManager
+ * \brief CLE memory manager implemented using OpenCL. Required by CLEImage class.
+ *
+ * This class serves as a base class for CLEsperanto data container for CLEImage class,
+ * which is similar to ImageBase class for Image class. However, all the image-related
+ * meta data will be already stored in image class (parent of CLEImage), therefore
+ * we did not name it CLEImageBase. Rather, this class is a CLE-specific data manager
+ * that provides functionalities for CPU-GPU data synchronization and grafting GPU data.
+ *
+ * \ingroup ITKCLE
```
Comment (thewtex) [inline include/itkCLEDataManager.h:40]: CLEsperanto

File: include/itkCLEDataManager.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ */
+template <typename TPixel, unsigned int VDim>
+class CLEDataManager : public Object // DataObject//
+{
+  template<typename,unsigned int> friend class CLEImage;
+
+public:
+  ITK_DISALLOW_COPY_AND_MOVE(CLEDataManager);
+
+  using Self = CLEDataManager;
```
Comment (thewtex) [inline include/itkCLEDataManager.h:50]: `Self` should be `SelfTemplate`

File: include/itkCLEDataManager.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  SetImagePointer(CLEImage<TPixel,VDim>* img)
+  {
+    m_Image = img;
+  }
+
+  CLEImage<TPixel, VDim>*
+  GetImagePointer()
+  {
+    return this->m_Image.GetPointer();
+  }
```
Comment (thewtex) [inline include/itkCLEDataManager.h:71]: Move to an `Image` specific class: `itk::CLEImageDataManager`.

File: include/itkCLEDataManager.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  //IsCPUBufferDirty() const
+  //{
+  //  return m_IsCPUBufferDirty;
+  //}
+  //
+  //bool
+  //IsGPUBufferDirty() const
+  //{
+  //  return m_IsGPUBufferDirty;
+  //}
```
Comment (thewtex) [inline include/itkCLEDataManager.h:115]: We want to exercise these in the test:

```
instance->SetGPUBufferDirty();
instance->Update();
instance->SetCPUBufferDirty();
instance->Update();
```

File: include/itkCLEDataManager.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  Graft(const CLEDataManager * data);
+
+protected:
+  CLEDataManager();
+  ~CLEDataManager() override;
+  void
+  PrintSelf(std::ostream & os, Indent indent) const override;
+
+private:
+  WeakPointer<CLEImage<TPixel, VDim>> m_Image; // WeakPointer has to be used here
```
Comment (thewtex) [inline include/itkCLEDataManager.h:140]: -> Image specific class

File: include/itkCLEDataManager.h [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  std::array<size_t, 3> m_BufferShape;
+  cle::Object m_GPUBuffer;
+  cle::Clesperanto cle;
+
+  /** checks if buffer needs to be updated */
+  bool m_IsGPUBufferDirty;
+  bool m_IsCPUBufferDirty;
+
+  /** Mutex lock to prevent r/w hazard for multithreaded code */
+  std::mutex m_Mutex;
```
Comment (thewtex) [inline include/itkCLEDataManager.h:151]: Remove if possible

File: include/itkCLEDataManager.h [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  ~CLEDataManager() override;
+  void
+  PrintSelf(std::ostream & os, Indent indent) const override;
+
+private:
+  WeakPointer<CLEImage<TPixel, VDim>> m_Image; // WeakPointer has to be used here
+                                               // to avoid SmartPointer loop
+  std::array<size_t, 3> m_BufferShape;
+  cle::Object m_GPUBuffer;
+  cle::Clesperanto cle;
```
Comment (thewtex) [inline include/itkCLEDataManager.h:144]: `static`  (at least for now). CLContextManager that holds a static singleton (can create an in issue if needed for the next PR.

File: include/itkCLEDataManager.hxx [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+template <typename TPixel, unsigned int VDim>
+void
+CLEDataManager<TPixel, VDim>::Allocate()
+{
+  auto size = m_Image->GetLargestPossibleRegion().GetSize();
+  m_BufferShape = { size[0],size[1],VDim==3?3:1};
+}
+
+template <typename TPixel, unsigned int VDim>
```
Comment (thewtex) [inline include/itkCLEDataManager.hxx:50]: `VDim` -> `VDimension`

File: include/itkCLEImage.h [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+  GetPixelContainer()
+  {
+    m_DataManager->SetGPUBufferDirty();
+    return Superclass::GetPixelContainer();
+  }
+
+  const PixelContainer*
+  GetPixelContainer() const
+  {
+    //m_DataManager->UpdateCPUBuffer();
```
Comment (thewtex) [inline include/itkCLEImage.h:127]: todo, etc

File: include/itkCLEImage.h [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+      OverrideImageTypeMacro(unsigned char, 3);
+      OverrideImageTypeMacro(signed char, 3);
+      OverrideImageTypeMacro(int, 3);
+      OverrideImageTypeMacro(unsigned int, 3);
+      OverrideImageTypeMacro(float, 3);
+      OverrideImageTypeMacro(double, 3);
+    }
+    */
+  }
+};
```
Comment (thewtex) [inline include/itkCLEImage.h:274]: Check into `itk.Image.{h,hxx}`: see if it does not have any of this content. If so, we may not need it.

File: test/itkCLEImageTest.cxx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+#include "itkCLEImage.h"
+
+#include "itkTestingMacros.h"
+#include "itkMath.h"
+
+using ItkImage1f = itk::CLEImage<float, 2>;
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:24]: inside the function

File: test/itkCLEImageTest.cxx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  }
+  unsigned int width, height;
+
+  width = 256;
+  height = 256;
+
+  // create CLEImages
+
+  // set size & region
+  ItkImage1f::Pointer srcA, srcB;
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:41]: Separate lines.

```
using ImageType = itk::CLEImage<float, 2>;
auto srcA = ImageType::New();
auto srcB = ImageType::New();

File: test/itkCLEImageTest.cxx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  size[1] = height;
+  ItkImage1f::RegionType region;
+  region.SetSize(size);
+  region.SetIndex(start);
+
+  // create
+  srcA = ItkImage1f::New();
+  srcA->SetRegions(region);
+  srcA->Allocate();
+  srcA->FillBuffer(1.0f);
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:57]: GetPixel
SetPixel

File: test/itkCLEImageTest.cxx [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  srcB = ItkImage1f::New();
+  srcB->SetRegions(region);
+  srcB->Allocate();
+  srcB->FillBuffer(1.0f);
+
+  if (*srcA != *srcB)
+    return EXIT_FAILURE;
+
+  // Push to the GPU
+  srcB->UpdateBuffers();
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:68]: Previous comment

File: test/itkCLEImageTest.cxx [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  srcA->Allocate();
+  srcA->FillBuffer(1.0f);
+
+  srcB = ItkImage1f::New();
+  srcB->SetRegions(region);
+  srcB->Allocate();
+  srcB->FillBuffer(1.0f);
+
+  if (*srcA != *srcB)
+    return EXIT_FAILURE;
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:65]: itkTestingMacros.h

**Comment by aktech** [comment]
The GPU tests didn't run because it seems the credentials you have in Cirun doesn't have permissions for the given AMI:

See error here: https://github.com/InsightSoftwareConsortium/ITKCLEsperanto/pull/21/checks?check_run_id=7986383145

File: example/ImageReadWrite.cxx [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  using PixelType = unsigned short;
+  constexpr unsigned int Dimension = 2;
+  using ImageType = itk::CLEImage<PixelType, Dimension>;
+ 
+  using ReaderType = itk::ImageFileReader<ImageType>;
+  using WriterType = itk::ImageFileWriter<ImageType>;
+
+  auto reader = ReaderType::New();
```
Comment (thewtex) [inline example/ImageReadWrite.cxx:25]: Better to use `itk::ReadImage`, `itk::WriteImage`:

https://examples.itk.org/src/io/imagebase/readanimage/documentation

https://examples.itk.org/src/io/imagebase/writeanimage/documentation

File: include/itkCLEContextManager.h [hunk 17]
```diff
@@ -15,22 +15,30 @@
...
-  generator->Initialize( 324 );
+/** \class CLEContextManager
+ *
+ * \brief Class to store the CLE context.
+ *
+ *  Write more documentation here...
+ *
+ * \ingroup CLEsperanto
+ */
+class CLEContextManager : public LightObject
```
Comment (thewtex) [inline include/itkCLEContextManager.h:33]: :+1:

File: include/itkCLEDataManager.hxx [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+template <typename TPixel, unsigned int VImageDimension>
+void
+CLEDataManager<TPixel, VImageDimension>::PrintSelf(std::ostream & os, Indent indent) const
+{
+  //os << indent << "CLEDataManager (" << this << ")" << std::endl;
+  //os << indent << "m_BufferSize: " << m_BufferSize << std::endl;
+  //os << indent << "m_IsGPUBufferDirty: " << m_IsGPUBufferDirty << std::endl;
+  //os << indent << "m_GPUBuffer: " << m_GPUBuffer << std::endl;
+  //os << indent << "m_IsCPUBufferDirty: " << m_IsCPUBufferDirty << std::endl;
+  //os << indent << "m_CPUBuffer: " << m_CPUBuffer << std::endl;
```
Comment (thewtex) [inline include/itkCLEDataManager.hxx:116]: uncomment

We can exercise these and more with

```
ITK_EXERCISE_BASIC_OBJECT_METHODS
```

from *itkTestingMacros.h*.

File: include/itkCLEImageDataManager.hxx [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+
+template <typename TPixel, unsigned int VImageDimension>
+void
+CLEImageDataManager<TPixel, VImageDimension>::UpdateGPUBuffer()
+{
+  if (m_IsGPUBufferDirty)
+  {
+    // Transfer itk::Image into our STL Buffer
+    std::vector<TPixel> buffer;
+    itk::ImageRegionConstIterator<CLEImage<TPixel, VImageDimension>> itr(m_Image, m_Image->GetLargestPossibleRegion());
```
Comment (thewtex) [inline include/itkCLEImageDataManager.hxx:74]: Largest -> Buffered

**Review by thewtex — CHANGES_REQUESTED** [review]
Good progress

**Comment by thewtex** [comment]
> The GPU tests didn't run because it seems the credentials you have in Cirun doesn't have permissions for the given AMI:

See error here: https://github.com/InsightSoftwareConsortium/ITKCLEsperanto/pull/21/checks?check_run_id=7986383145

@aktech thanks for the note!

@tbirdso could you please take a look?

**Review by thewtex — CHANGES_REQUESTED** [review]
Additional todos:

- [ ] test `Graft`, as with the [itkImageTest](https://github.com/InsightSoftwareConsortium/ITK/blob/05e9b3a16b52e9f34609a07b70c53f663468ab76/Modules/Core/Common/test/itkImageTest.cxx#L229-L235)
- [x] test equality after modifications [with the comparison image filter](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples/blob/ffdabca2f41e91a062b6b616f300f55ef088eddc/src/ImageCompareCommand.cxx#L284-L293) 
```
  using ComparisonFilterType = ComparisonImageFilter< CLEIImageType,
CLEImageType >;
```
- [ ] Addresss inline comments
- [ ] CI tests passing across platforms

File: include/itkCLEDataManager.hxx [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+CLEDataManager<TPixel, VImageDimension>::PrintSelf(std::ostream & os, Indent indent) const
+{
+  os << indent << "CLEDataManager (" << this << ")" << std::endl;
+  os << indent << "m_BufferShape: ("
+    << m_BufferShape[0] << ", "
+    << m_BufferShape[1] << ", "
+    << m_BufferShape[2] << ")" << std::endl;
+  os << indent << "m_IsGPUBufferDirty: " << m_IsGPUBufferDirty << std::endl;
+  os << indent << "m_IsCPUBufferDirty: " << m_IsCPUBufferDirty << std::endl;
+  //os << indent << "m_GPUBuffer: " << m_GPUBuffer << std::endl;
```
Comment (thewtex) [inline include/itkCLEDataManager.hxx:102]: Add

File: include/itkCLEImageDataManager.hxx [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    std::vector<TPixel> buffer = CLEContextManager::GetInstance().Pull<TPixel>(m_GPUBuffer);
+    size_t i = 0;
+    itk::ImageRegionIterator<CLEImage<TPixel, VImageDimension>> itr(m_Image, m_Image->GetBufferedRegion());
+    itr.GoToBegin();
+    while (!itr.IsAtEnd())
+    {
+      itr.Set(buffer[i++]);
+      ++itr;
+    }
+    std::cout << "I am pulling from the GPU\n";
```
Comment (thewtex) [inline include/itkCLEImageDataManager.hxx:62]: Use `itkDebugMacro`, then set `DebugOn` in the test.

File: include/itkCLEImageDataManager.hxx [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    itk::ImageRegionConstIterator<CLEImage<TPixel, VImageDimension>> itr(m_Image, m_Image->GetLargestPossibleRegion());
+    itr.GoToBegin();
+    while (!itr.IsAtEnd())
+    {
+      buffer.push_back(itr.Get());
+      ++itr;
+    }
+    m_GPUBuffer.Reset();
+    m_GPUBuffer = CLEContextManager::GetInstance().Push<TPixel>(buffer, m_BufferShape);
+    std::cout << "I am pushing to the GPU\n";
```
Comment (thewtex) [inline include/itkCLEImageDataManager.hxx:84]: Use `itkDebugMacro`, then set `DebugOn` in the test.

File: test/itkCLEImageTest.cxx [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+  if (argc > 1)
+  {
+    std::cout << "received " << argc << " arguments, but didn't expect any."
+      << "first ignored argument: " << argv[1] << std::endl;
+  }
+  unsigned int width, height;
+  using ImageType = itk::CLEImage<float, 2>;
+
+  auto cleImage = ImageType::New();
+  ITK_EXERCISE_BASIC_OBJECT_METHODS(cleImage, CLEImage, Image)
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:81]: :+1:

File: test/itkCLEImageTest.cxx [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+  srcA->SetRegions(region);
+  srcA->Allocate();
+  srcB->SetRegions(region);
+  srcB->Allocate();
+
+
+  // Testing FillBuffer
+  srcA->FillBuffer(1.0f);
+  srcB->FillBuffer(1.0f);
+  if (Compare<ImageType>(srcA,srcB))
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:110]: :+1:

**Review by thewtex — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 25]
```diff
@@ -45,18 +45,21 @@ OrganizeTargets(${clci_SOURCE_DIR} clci)
...
-# Create the module by processing itk-module.cmake
-include(ITKModuleExternal)
+  # Create the module by processing itk-module.cmake
+  include(ITKModuleExternal)
 else()
   set(ITK_DIR ${CMAKE_BINARY_DIR})
   itk_module_impl()
 endif()
 
+add_subdirectory(example)
```
Comment (thewtex) [inline CMakeLists.txt:63]: This is failing CI, does not appear to be a usable example.

Let's remove for now, investigate adding Jupyter notebooks later.

File: include/itkCLEImage.h [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+   */
+  void
+  DataHasBeenGenerated() override
+  {
+    Superclass::DataHasBeenGenerated();
+    //if (m_DataManager->IsCPUBufferDirty())
+    {
+      m_DataManager->Modified();
+    }
+  }
```
Comment (thewtex) [inline include/itkCLEImage.h:204]: Add `PrintSelf`.

File: include/itkCLEImage.h [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+  using ConstWeakPointer = WeakPointer<const Self>;
+
+  using ImageDataManager = CLEImageDataManager<TPixel, VImageDimension>;
+
+  itkNewMacro(Self);
+
+  itkTypeMacro(CLEImage, Image);
+
+  itkGetModifiableObjectMacro(DataManager, ImageDataManager);
+  CLEDataManager<TPixel,VImageDimension>*
```
Comment (thewtex) [inline include/itkCLEImage.h:59]: Use the typedef.

File: include/itkCLEImage.h [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+    m_DataManager->SetGPUBufferDirty();
+    return Superclass::GetPixelAccessor();
+  }
+
+  /** Return the Pixel Accesor object */
+  const AccessorType
+  GetPixelAccessor() const
+  {
+    return Superclass::GetPixelAccessor();
+  }
```
Comment (thewtex) [inline include/itkCLEImage.h:172]: same as base -- remove

File: include/itkCLEImage.h [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+  {
+    m_DataManager->SetGPUBufferDirty();
+    return NeighborhoodAccessorFunctorType();
+  }
+
+  const NeighborhoodAccessorFunctorType
+  GetNeighborhoodAccessor() const
+  {
+    return NeighborhoodAccessorFunctorType();
+  }
```
Comment (thewtex) [inline include/itkCLEImage.h:187]: same as base -- remove

File: include/itkCLEImage.h [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+      m_DataManager->Modified();
+    }
+  }
+
+protected:
+  void
+  Graft(const DataObject * data) override;
+  CLEImage();
+  ~CLEImage() override;
+  using Superclass::Graft;
```
Comment (thewtex) [inline include/itkCLEImage.h:211]: move above `Graft` on line 208

File: include/itkCLEImage.hxx [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+#ifndef itkCLEImage_hxx
+#define itkCLEImage_hxx
+
+
+namespace itk
+{
+template <typename TPixel, unsigned int VImageDimension>
+CLEImage<TPixel, VImageDimension>::CLEImage()
+{
+  m_DataManager = CLEImageDataManager<TPixel, VImageDimension>::New();
```
Comment (thewtex) [inline include/itkCLEImage.hxx:27]: use typedef

File: include/itkCLEImageDataManager.hxx [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+#include "itkCLEContextManager.h"
+
+namespace itk
+{
+
+template <typename TPixel, unsigned int VImageDimension>
+void
+CLEImageDataManager<TPixel, VImageDimension>::Allocate()
+{
+  auto size = m_Image->GetLargestPossibleRegion().GetSize();
```
Comment (thewtex) [inline include/itkCLEImageDataManager.hxx:30]: `BufferedRegion`

File: include/itkCLEDataManager.h [hunk 33]
```diff
@@ -0,0 +1,125 @@
...
+  /** actual GPU->CPU memory copy takes place here */
+  virtual void
+  UpdateCPUBuffer();
+
+  /** actual CPU->GPU memory copy takes place here */
+  virtual void
+  UpdateGPUBuffer();
+
+  std::array<size_t, 3> m_BufferShape;
+  cle::Object m_GPUBuffer;
```
Comment (thewtex) [inline include/itkCLEDataManager.h:112]: -> `unique_ptr<cle::Object>` or `cli::Object *`

File: test/itkCLEImageTest.cxx [hunk 34]
```diff
@@ -0,0 +1,161 @@
...
+
+  // Testing FillBuffer
+  srcA->FillBuffer(1.0f);
+  srcB->FillBuffer(1.0f);
+  if (Compare<ImageType>(cleImage,srcB))
+    return EXIT_FAILURE;
+
+  // Testing SetPixel
+  srcA->SetPixel(idx, 7.0f);
+  srcB->SetPixel(idx, 7.0f);
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:127]: Test `GetPixel`

File: test/itkCLEImageTest.cxx [hunk 35]
```diff
@@ -0,0 +1,161 @@
...
+  auto pxlCtrA = srcA->GetPixelContainer();
+  auto pxlCtrB = srcB->GetPixelContainer();
+  (*pxlCtrA)[7] = 55.0f;
+  (*pxlCtrB)[7] = 55.0f;
+  if (Compare<ImageType>(srcA, srcB))
+    return EXIT_FAILURE;
+
+  // Testing SetPixelContainer
+  // Testing GetPixelAccessor
+  // Testing GetNeighborhoodAccessor
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:157]: Remove

**Review by thewtex — CHANGES_REQUESTED** [review]

File: test/itkCLEImageTest.cxx [hunk 36]
```diff
@@ -0,0 +1,181 @@
...
+  ITK_TEST_EXPECT_EQUAL(A, B);
+
+  // Testing GetBufferPointer
+  auto bufptrA = srcA->GetBufferPointer();
+  auto bufptrB = srcB->GetBufferPointer();
+  bufptrA[5] = 12.0f;
+  bufptrB[5] = 12.0f;
+  A = srcA->GetPixel(idx);
+  B = srcB->GetPixel(idx);
+  ITK_TEST_EXPECT_EQUAL(A, B);
```
Comment (thewtex) [inline test/itkCLEImageTest.cxx:159]: In these tests, `B` should explicitly be `12.0f` in the case, etc. `A` and `B` could both be `0.0` if setting the values failed and the test would still pass.

File: include/itkCLEImageDataManager.hxx [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+#ifndef itkGPUImageDataManager_hxx
+#define itkGPUImageDataManager_hxx
```
Comment (thewtex) [inline include/itkCLEImageDataManager.hxx:20]: update macro name

**Review by thewtex — APPROVED** [review]

**Pull request merged**